vllm/core/scheduler.py

```diff
@@ -336,76 +300,243 @@ def _schedule(self) -> SchedulerOutputs:
+            # exceed the maximum number of sequences.
+            num_new_seqs = seq_group.get_max_num_running_seqs()
+            if (num_curr_seqs + num_new_seqs >
+                    self.scheduler_config.max_num_seqs):
+                break
+
+            if curr_loras is not None and lora_int_id > 0:
+                curr_loras.add(lora_int_id)
+            self.waiting.popleft()
+            self._allocate(seq_group)
+            seq_groups.append(
+                ScheduledSequenceGroup(seq_group=seq_group,
+                                       token_chunk_size=num_prompt_tokens))
+            num_batched_tokens += num_prompt_tokens
+            num_curr_seqs += num_new_seqs
+
+        self.waiting.extendleft(leftover_waiting_sequences)
+        if len(seq_groups) > 0:
+            self.prev_prompt = True
+
+        return SchedulerPrefillOutputs(seq_groups=seq_groups,
+                                       ignored_seq_groups=ignored_seq_groups,
+                                       num_batched_tokens=num_batched_tokens)
 
-            self.swapped.extendleft(leftover_swapped)
+    def _schedule(self) -> SchedulerOutputs:
+        """Batch requests that are queued..
+        
+        The current policy is designed to opimimize the throughput. First,
+        it batches as many prefill requests as possible. And it schedules
+        decodes. If there's a pressure on GPU memory, decode requests can
+        be swapped or preempted.
+        """
+        token_budget = self.scheduler_config.max_num_batched_tokens
+
+        if self.swapped:
+            # If requests are swapped, prefill cannot be scheduled because
+            # it means there's no additional available blocks.
+            prefills = SchedulerPrefillOutputs.create_empty()
+        else:
+            # First schedule as many prefilling requests as possible,
+            # then schedule decoding requests.
+            prefills = self._schedule_prefills(token_budget)
+        token_budget -= prefills.num_batched_tokens
+
+        # Don't schedule decodes if prefills are scheduled.
+        # This policy is to maximize the throughput.
+        if len(prefills.seq_groups) > 0:
+            decodes = SchedulerDecodeOutputs.create_empty()
+        else:
+            decodes = self._schedule_decodes(token_budget)
+        token_budget -= decodes.num_batched_tokens
 
-        # Each sequence in the generation phase only takes one token slot.
-        # Therefore, the number of batched tokens is equal to the number of
-        # sequences in the RUNNING state.
-        num_batched_tokens = sum(
-            seq_group.num_seqs(status=SequenceStatus.RUNNING)
-            for seq_group in self.running)
+        # If any sequence group is preempted, do not swap in any sequence group.
+        if decodes.num_preempted_seqs > 0:
+            swapped_in = SchedulerSwappedOutputs.create_empty()
+        else:
+            swapped_in = self._schedule_swapped()
```

Comment:
(not for this PR) swapped should also consume token budget

Reply:
yeah, I am just following the same behavior as master now